main.py

```diff
@@ -14,29 +14,97 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 
+from typing import Optional
```

Comment:
You've imported this and haven't used it, you've also said in your opening comment "Added type hints through out the codebase" when you've not done that.